### PR TITLE
Modernise C# style within System.Linq

### PR DESCRIPTION
**Pull request by JonHanna**
In particular, `is`-with patterns is applicable quite a few places.

### CODE REVIEW

File: src/System.Linq/src/System/Linq/First.cs [hunk 1]
```diff
@@ -34,17 +32,11 @@ public static TSource First<TSource>(this IEnumerable<TSource> source, Func<TSou
             return first;
         }
 
-        public static TSource FirstOrDefault<TSource>(this IEnumerable<TSource> source)
-        {
-            bool found;
-            return source.TryGetFirst(out found);
-        }
+        public static TSource FirstOrDefault<TSource>(this IEnumerable<TSource> source) =>
```
Comment (sharwell) [inline src/System.Linq/src/System/Linq/First.cs]: :memo: I find this hard to read. It's much clearer when the `=>` appears on the following line.

File: src/System.Linq/src/System/Linq/Take.cs [hunk 2]
```diff
@@ -21,14 +21,12 @@ public static IEnumerable<TSource> Take<TSource>(this IEnumerable<TSource> sourc
                 return EmptyPartition<TSource>.Instance;
             }
 
-            IPartition<TSource> partition = source as IPartition<TSource>;
-            if (partition != null)
+            if (source is IPartition<TSource> partition)
```
Comment (sharwell) [inline src/System.Linq/src/System/Linq/Take.cs]: 💭 Seems like some of these cases could benefit from the `switch` syntax (e.g. more than one special type consideration, and in every case the operation just returns something):

```csharp
switch (source)
{
case IPartition<TSource> partition:
  return partition.Take(count);

case IList<TSource> sourceList:
  return new ListPartition<TSource>(sourceList, 0, count - 1);

default:
  return new EnumerablePartition<TSource>(source, 0, count - 1);
}
```

File: src/System.Linq/src/System/Linq/First.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/First.cs]: And I find it harder to read when the => appears on the following line ;)

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/First.cs]: I agree with @sharwell on this, but have been following @stephentoub on it after an earlier discussion on the same point.

File: src/System.Linq/src/System/Linq/Take.cs (hunk 2 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Take.cs]: Good idea.

File: src/System.Linq/src/System/Linq/First.cs (hunk 1 above)
Comment (jnm2) [inline src/System.Linq/src/System/Linq/First.cs]: `=>` on the following line messes up the relative following indent.

File: src/System.Linq/src/System/Linq/First.cs [hunk 3]
```diff
@@ -34,17 +32,11 @@ public static TSource First<TSource>(this IEnumerable<TSource> source, Func<TSou
...
+            source.TryGetFirst(out bool found);
```
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/First.cs]: I quite like naming unused out variables like this `_` to reflect their unimportance to the given use. If other people also find this useful I could apply that here, but I don't recall anything suggesting this in the style guide.

Comment (jnm2) [inline src/System.Linq/src/System/Linq/First.cs]: What about using the discard `out _`?

File: src/System.Linq/src/System/Linq/Take.cs (hunk 2 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Take.cs]: That seems quite nice (while most of this was inspired by returning to the code after a break and seeing things I'm now used to see done differently, I don't have the same reflexes for pattern `switch` yet). It could be even nicer if I could change which argument is complained about when more than one would trigger an ANE, but that's a very slightly breaking change, so not really worth it.

File: src/System.Linq/src/System/Linq/First.cs (hunk 3 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/First.cs]: Actually, I just remembered that this is now actually encouraged by C#, so I shall when I come back to this, this evening.

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/First.cs]: @jnm2 yes, you're referring to exactly the moment of esprit d'escalier I had :)

File: src/System.Linq/src/System/Linq/Take.cs (hunk 2 above)
Comment (sharwell) [inline src/System.Linq/src/System/Linq/Take.cs]: Technically the language would allow this here, but I'm not sure it's more readable.

```csharp
switch (source)
{
case null:
  throw Error.ArgumentNull(nameof(source));

case var _ when (count <= 0):
  return EmptyPartition<TSource>.Instance;

case IPartition<TSource> partition:
  return partition.Take(count);

case IList<TSource> sourceList:
  return new ListPartition<TSource>(sourceList, 0, count - 1);

default:
  return new EnumerablePartition<TSource>(source, 0, count - 1);
}
```

I lean towards staying with `if` statements at least for argument validation.

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Take.cs]: Switching on `_` seems nasty to me.

File: src/System.Linq/src/System/Linq/First.cs (hunk 3 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/First.cs]: Hmm. The compiler is complaining about discards. Is this something where current compilers differ?

**Comment by JonHanna** [comment]
Test NETFX x86 Release Build
(File-related tests failed. Have we changed how failures are reported because I couldn't find the actual test failures).

**Comment by JonHanna** [comment]
Run Linux x64 Release Build

(java.lang.NullPointerException in jenkins).

**Comment by JonHanna** [comment]
Run Linux x64 Release Build
(Caused: java.lang.IllegalArgumentException: Failed to prepare archiveArtifacts step)

**Comment by JonHanna** [comment]
Test Linux x64 Release Build

File: src/System.Linq/src/System/Linq/First.cs (hunk 3 above)
Comment (jnm2) [inline src/System.Linq/src/System/Linq/First.cs]: What's the error message?

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/First.cs]: The same "The name _ does not exist in the current…" you'd get with C#6

File: src/System.Linq/src/System/Linq/Zip.cs [hunk 4]
```diff
@@ -8,6 +8,7 @@ namespace System.Linq
 {
     public static partial class Enumerable
     {
+
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Zip.cs]: Why?

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Zip.cs]: Yuck :(

**Comment by stephentoub** [comment]
Have you compared the IL for the old and the new versions, in particular for the ones that use pattern matching?  I just took a quick look at one case (TryGetCount), and the new IL looks worse, e.g. 20% bigger.

**Comment by JonHanna** [comment]
Interesting. I'd compared the difference with `is` patterns before and found that the IL is equivalent or slightly better that way (very close to how `?.` tends to be equivalent or slightly better for reference types, though its worse if the target type is a non-nullable value type). But I didn't compare between that and the switch form.

The main differences seem to be:

1. Short-circuiting on `null`, which is quite clever generally, but not useful here when we know the value isn't null.
2. Evaluating the expression `source` and storing the result. Obviously necessary if the expression was something other than a bare reference to an argument, but not useful here. I can't see how it protects against an observable change unless perhaps there was a mutable `when` expression that overwrote `source` in which case it would be semantically necessary.

In any case, I think I shall undo that commit and revert back to just changing to `is` patterns.

**Comment by JonHanna** [comment]
I left the `switch` in `ReverseIterator.GetCount()` because there it was an improvement in the IL in reducing the number of field accesses.

**Comment by JonHanna** [comment]
Test Linux x64 Release Build

File: src/System.Linq/src/System/Linq/Concat.cs [hunk 5]
```diff
@@ -61,16 +60,15 @@ internal Concat2Iterator(IEnumerable<TSource> first, IEnumerable<TSource> second
 
             internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
             {
-                bool hasOnlyCollections = _first is ICollection<TSource> &&
-                                          _second is ICollection<TSource> &&
-                                          next is ICollection<TSource>;
+                bool hasOnlyCollections = next is ICollection<TSource> &&
+                                          _first is ICollection<TSource> &&
+                                          _second is ICollection<TSource>;
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Concat.cs]: Why did you change the order here?

File: src/System.Linq/src/System/Linq/Concat.cs [hunk 6]
```diff
@@ -61,16 +60,15 @@ internal Concat2Iterator(IEnumerable<TSource> first, IEnumerable<TSource> second
...
                 return new ConcatNIterator<TSource>(this, next, 2, hasOnlyCollections);
             }
 
             public override int GetCount(bool onlyIfCheap)
             {
-                int firstCount, secondCount;
-                if (!EnumerableHelpers.TryGetCount(_first, out firstCount))
+                if (!EnumerableHelpers.TryGetCount(_first, out int firstCount))
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Concat.cs]: I personally find out declarations like this difficult.  It's one thing if the declared variable is only used in the scope of the body of the if, or if the method is small, but otherwise I find it makes it difficult when later reading the method to figure out where the variable came from... I can no longer just look quickly up the left-hand side of the code and at the method signature; I instead need to essentially scan the full contents of many lines looking to see where it was defined.  I'd prefer we not use them in situations like this.

File: src/System.Linq/src/System/Linq/Cast.cs [hunk 7]
```diff
@@ -30,21 +30,8 @@ private static IEnumerable<TResult> OfTypeIterator<TResult>(IEnumerable source)
...
-
-            if (source == null)
-            {
-                throw Error.ArgumentNull(nameof(source));
-            }
-
-            return CastIterator<TResult>(source);
-        }
+        public static IEnumerable<TResult> Cast<TResult>(this IEnumerable source) =>
+            source as IEnumerable<TResult> ?? CastIterator<TResult>(source ?? throw Error.ArgumentNull(nameof(source)));
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Cast.cs]: Certainly concise, but IMO harder to understand than the original.

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Cast.cs]: I to-and-fro'd on this myself, so I'm not upset to revert it.

File: src/System.Linq/src/System/Linq/Concat.cs (hunk 6 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Concat.cs]: NP, though personally I find the new style an improvement because it's clearly assigned at that point.

File: src/System.Linq/src/System/Linq/Concat.cs (hunk 5 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Concat.cs]: TBH, I thought I'd undone that as not relevant to what's being done here, but clearly I'd committed it. It seems reasonable that if we're going to do something on two fields and a local and any of them can short-circuit away from the rest to operate on the local first.

**Comment by stephentoub** [comment]
Thanks, @JonHanna.
